lla_to_ecef.cpp

```diff
@@ -0,0 +1,390 @@
+                        break;
+                    case 2:
+                        lla_entry.longitude = std::stod(value);
+                        break;
+                    case 3:
+                        lla_entry.altitude = std::stod(value) * KM_TO_M;
+                    default:
+                        break;
+                }
+                index++;
+            }
+
+            lla_data.push_back(lla_entry);
+        }
+
+        file.close();
+    } else {
+        throw std::runtime_error(std::format(
+        "Could not open file {}. Exiting", filename));
+    }
+
+    return lla_data;
+}
+
+/**
+ * get_ecef_vel_at_poi
+ * @brief Calculates ECEF velocity at a point of interest.
+ *
+ * Linear interpolation formula is derived by solving for the unkown y-value
+ * from the equation of slopes:
+ *
+ *          slope = (y - y0) / (x - x0) = (y1 - y0) / (x1 - x0)
+ *
+ * @param lla_data          timestamped LLA data (deg, deg, m)
+ * @param point_of_interest timespec for point of interest
+ * @param print_output      flag to print output to stdout
+ *
+ * @return XYZ ECEF velocity at the point of interest
+ */
+std::vector<double> get_ecef_vel_at_poi(
+    const std::vector<PositionLLA>& lla_data,
+    const timespec& point_of_interest,
+    const bool& print_output) {
+
+  // Search data for point of interest
+  std::vector<int> search_results =
+    find_match_or_nearest(lla_data, point_of_interest);
+
+  std::vector<double> velocity_result;
+
+  // If no exact time match for point of interest was found, interpolate
+  if (search_results[1] == INDEX_ERR) {
+    if (print_output) {
+        std::cout << "No exact match found for timestamp: ";
+        std::cout << point_of_interest.tv_sec << " s, ";
+        std::cout << point_of_interest.tv_nsec << " ns";
+        std::cout << ". Interpolating from nearby points." << std::endl;
+    }
+
+    // Get nearby points in ECEF
+    PositionVelocityECEF before_before = lla_to_ecef_pos(
+        lla_data[search_results[0] - 1],
+        WGS84Params::A, WGS84Params::B, WGS84Params::E);
+    PositionVelocityECEF before = lla_to_ecef_pos(
+        lla_data[search_results[0]],
+        WGS84Params::A, WGS84Params::B, WGS84Params::E);
+    PositionVelocityECEF after = lla_to_ecef_pos(
+        lla_data[search_results[2]],
+        WGS84Params::A, WGS84Params::B, WGS84Params::E);
+
+    // Calculate velocity at nearby points
+    get_vel_ecef(before_before, &before);
+    get_vel_ecef(before, &after);
+
+    // interpolate_2d to find velocity at point of interest
+    velocity_result = interpolate_ecef_vel(before, after, point_of_interest);
+
+  } else {
+    if (print_output) {
+        std::cout << "Exact match found for timestamp: ";
+        std::cout << point_of_interest.tv_sec << " s, ";
+        std::cout << point_of_interest.tv_nsec << " ns" << std::endl;
+    }
+
+    PositionVelocityECEF pv_ecef_before_poi = lla_to_ecef_pos(
```

Comment:
nit: rather than duplicating all this code, I would have the if-else set some intermediary argument variables that are then passed to the function calls.

Reply:
agreed